### PR TITLE
OKTA-89999: Better label for remember device

### PR DESCRIPTION
**Pull request by mauriciocastillosilva-okta**
Remember Device use to say "Trust this device", but now instead it will convey a more meaningful text
- Do not challenge me on this device again
- Do not challenge me on this device for the next 15 minutes

Note: we won't show it if the server is set to prompt for mfa always

In addition I am removing the device cookie since it is no longer needed. The server will handle the logic for that

Screencast: https://okta.box.com/s/agm58j63aywxclop681r8lvqy4cvh2uw

@srinivasanagandla-okta can you take a look at FactorUtil.js please? This is the logic to figure out if we check or not the checkbox by default

### CODE REVIEW

File: src/MfaVerifyController.js [hunk 1]
```diff
@@ -89,12 +89,12 @@ function (Okta, Checkbox, BaseLoginController, CookieUtil, TOTPForm, YubikeyForm
         }
 
         // Remember Device checkbox resides outside of the Push and TOTP forms.
-        if (this.settings.get('features.rememberDevice')) {
+        if (this.settings.get('features.rememberDevice') && this.options.appState.get('allowRememberDevice')) {
```
Comment (Ujjwalreddy-okta) [inline src/MfaVerifyController.js]: We will not need the rememberDevice config anymore, right? (Since all the logic resides in the API)

Comment (Ujjwalreddy-okta) [inline src/MfaVerifyController.js]: Also, since we are using this - `this.settings.get('features.rememberDevice') && this.options.appState.get('allowRememberDevice')` everywhere, it might make sense to make this a model attribute or a helper function and use it everywhere.

File: src/models/AppState.js [hunk 2]
```diff
@@ -55,6 +55,21 @@ function (Okta, Q, Factor, BrowserFeatures, Errors) {
...
   }
 
+  function getMinutesString(factorLifetimeInMinutes) {
+    if (factorLifetimeInMinutes > 60 && factorLifetimeInMinutes <= 1440) {
+      var lifetimeInHours = (factorLifetimeInMinutes / 60);
+      return Okta.loc('hours', 'login', [lifetimeInHours]);
+    } else if (factorLifetimeInMinutes > 1440 && factorLifetimeInMinutes < 2880) {
+      var lifetimeInDays = (factorLifetimeInMinutes / 1440);
+      return Okta.loc('days', 'login', [lifetimeInDays]);
+    }
```
Comment (Ujjwalreddy-okta) [inline src/models/AppState.js]: `60 < factorLifetimeInMinutes <=1440` and `1440 < factorLifetimeInMinutes < 2880` ?

File: src/models/AppState.js [hunk 3]
```diff
@@ -388,6 +407,30 @@ function (Okta, Q, Factor, BrowserFeatures, Errors) {
...
+        deps: ['policy'],
+        fn: function (policy) {
+          if (policy && policy.rememberDeviceLifetimeInMinutes > 0) {
+            var timeString = getMinutesString(policy.rememberDeviceLifetimeInMinutes);
+            return Okta.loc('rememberDevice.timebased', 'login', [timeString]);
+          } else if (policy && policy.rememberDeviceLifetimeInMinutes === 0) {
+            return Okta.loc('rememberDevice.devicebased', 'login');
+          }
+          return Okta.loc('rememberDevice', 'login');
+        }
```
Comment (Ujjwalreddy-okta) [inline src/models/AppState.js]: nit:
```javascript
if (!policy) {
  return Okta.loc('rememberDevice', 'login');
}

then if()/else{} ?
```

File: src/models/Factor.js [hunk 4]
```diff
@@ -185,13 +184,7 @@ function (Okta, Q, CookieUtil, factorUtil, BaseLoginModel) {
 
     save: function () {
       var rememberDevice = !!this.get('rememberDevice');
-      var username = this.appState.get('username');
       // Set/Remove the remember device cookie based on the remember device input.
-      if (rememberDevice) {
-        CookieUtil.setDeviceCookie(username);
-      } else {
-        CookieUtil.removeDeviceCookie();
-      }
```
Comment (Ujjwalreddy-okta) [inline src/models/Factor.js]: Since we don't use the device cookie anymore, we should also remove the device cookie methods from CookieUtil

File: src/util/FactorUtil.js [hunk 5]
```diff
@@ -141,20 +141,11 @@ define(['okta' , './CookieUtil'], function (Okta, CookieUtil) {
...
-      rememberDevice = true;
-    } else if (rememberDeviceUsername || username) {
-      rememberDevice = (rememberDeviceUsername === username);
-    }
-
-    return rememberDevice;
+    var rememberDeviceValueFromServer = appState && appState.get('rememberDeviceByDefault');
+    // rememberDevice is true if 'rememberDeviceAlways' is on (feature flag)
+    // or if the server returns true for this user/device
+    return rememberDeviceAlways || rememberDeviceValueFromServer;
```
Comment (Ujjwalreddy-okta) [inline src/util/FactorUtil.js]: We will not rememberDeviceAlways config too, since we get this information from the server.

File: src/models/Factor.js (hunk 4 above)
Comment (mauriciocastillosilva-okta) [inline src/models/Factor.js]: I was 99% sure I did that, but clearly the 1% won this time... sigh...
I'll do that

File: src/models/AppState.js (hunk 2 above)
Comment (mauriciocastillosilva-okta) [inline src/models/AppState.js]: I tried this and it does not evaluate the same, so I'll keep what I have

**Comment by mauriciocastillosilva-okta** [comment]
Feedback comments addresed (2nd commit) and I also updated the i18n version to the latest, take a look please @Ujjwalreddy-okta

File: src/models/Settings.js [hunk 6]
```diff
@@ -78,6 +78,8 @@ function (Okta, Errors, BrowserFeatures) {
       'helpLinks.custom': 'array'
     },
 
+    // extraProperties: /^features.(rememberDevice|rememberDeviceAlways)\./,
+
```
Comment (Ujjwalreddy-okta) [inline src/models/Settings.js]: Let's remove this commented out code.

Comment (mauriciocastillosilva-okta) [inline src/models/Settings.js]: my mistake, not sure how this got in. Will remove

File: test/spec/MfaVerify_spec.js [hunk 7]
```diff
@@ -241,21 +244,32 @@ function (Q, _, $, Duo, OktaAuth, LoginUtil, CryptoUtil, Util, MfaVerifyForm, Be
...
+          return setup(resMfaDevicePolicy).then(function (test) {
+            expect(test.form.rememberDeviceLabelText()).toEqual('Do not challenge me on this device again');
+          });
+        });
+        itp('has the right text for time based policy', function () {
+          return setup(resMfaTimePolicy).then(function (test) {
+            expect(test.form.rememberDeviceLabelText()).toEqual(
+              'Do not challenge me on this device for the next 15 minutes');
           });
         });
```
Comment (Ujjwalreddy-okta) [inline test/spec/MfaVerify_spec.js]: Can we also add two tests - one for 1 min and the other for hours/days to test getMinutesString() function?

File: test/spec/MfaVerify_spec.js [hunk 8]
```diff
@@ -1559,17 +1553,6 @@ function (Q, _, $, Duo, OktaAuth, LoginUtil, CryptoUtil, Util, MfaVerifyForm, Be
...
-          return setupDuo({ 'features.rememberDeviceAlways': true }).then(function () {
-            expect($.ajax.calls.count()).toBe(2);
-            Expect.isJsonPost($.ajax.calls.argsFor(1), {
-              url: 'https://foo.com/api/v1/authn/factors/ost947vv5GOSPjt9C0g4/verify?rememberDevice=true',
-              data: {
-                stateToken: 'testStateToken'
-              }
-            });
-          });
-        });
```
Comment (Ujjwalreddy-okta) [inline test/spec/MfaVerify_spec.js]: We still want to keep these tests since we are testing "rememberDevice=true" query param. The test just needs to be set up in a different way.

**Comment by Ujjwalreddy-okta** [comment]
Also, we want to remove these settings in Settings.js (since they will not be used anymore) -

```javascript
       'features.rememberDevice': ['boolean', true, true],
       'features.rememberDeviceAlways': ['boolean', true, false],
```

**Comment by Ujjwalreddy-okta** [comment]
Looks good overall, just a few final comments.

File: test/spec/MfaVerify_spec.js (hunk 7 above)
Comment (mauriciocastillosilva-okta) [inline test/spec/MfaVerify_spec.js]: done

File: test/spec/MfaVerify_spec.js (hunk 8 above)
Comment (mauriciocastillosilva-okta) [inline test/spec/MfaVerify_spec.js]: I added one back

File: src/VerifyDuoController.js [hunk 9]
```diff
@@ -90,14 +81,17 @@ function (Okta, Duo, Q, CookieUtil, FactorUtil, FormController, Enums, FormType,
         // jquery decides the Content-Type instead of it being a JSON type). Enroll/Verify DUO
         // are the only two places where we actually do this.
         // NOTE - If we ever decide to change this, we should test this very carefully.
+        var rememberDevice = this.get('rememberDevice');
         return Q($.post(url, data))
         .then(function () {
           return self.doTransaction(function(transaction) {
+            if (rememberDevice) {
+              transaction.data._links.next.href += '?rememberDevice=true';
+            }
```
Comment (mauriciocastillosilva-okta) [inline src/VerifyDuoController.js]: @Ujjwalreddy-okta this is the tricky part in duo controller. Take a look please.

Comment (Ujjwalreddy-okta) [inline src/VerifyDuoController.js]: Yeah, we don't access links and we should not know the next available links, the Auth SDK handles it. Let's change this, I can help you with this.

**Comment by Ujjwalreddy-okta** [comment]
Looks good. When @lboyette-okta makes the change to the poll() logic to take in data, we can do on final review.

File: src/VerifyDuoController.js [hunk 10]
```diff
@@ -111,11 +106,11 @@ function (Okta, Duo, Q, CookieUtil, FactorUtil, FormController, Enums, FormType,
...
       postRender: function () {
         this.add('<iframe frameborder="0"></iframe>');
-        if (this.settings.get('features.rememberDevice')) {
+        if (this.options.appState.get('allowRememberDevice')) {
           this.addInput({
             label: false,
             'label-top': true,
-            placeholder: Okta.loc('rememberDevice', 'login'),
+            placeholder: this.options.appState.get('rememberDeviceLabel'),
             className: 'margin-btm-0',
```
Comment (Ujjwalreddy-okta) [inline src/VerifyDuoController.js]: `this.appState` instead of `this.options.appState` to keep it consistent with the way we are using it above.

File: test/spec/LoginRouter_spec.js [hunk 11]
```diff
@@ -23,13 +23,14 @@ define([
...
   'helpers/xhr/ERROR_invalid_token',
+  'helpers/xhr/MFA_REQUIRED_policy_time_based',
   'util/Errors',
   'util/BrowserFeatures'
 ],
 function (Okta, Q, Backbone, xdomain, SharedUtil, CryptoUtil, CookieUtil, OktaAuth, Util, Expect, Router,
           $sandbox, PrimaryAuthForm, RecoveryForm, MfaVerifyForm, resSuccess, resRecovery,
           resMfa, resMfaRequiredDuo, resMfaRequiredOktaVerify, resMfaChallengeDuo, resMfaChallengePush,
-          errorInvalidToken, Errors, BrowserFeatures) {
+          errorInvalidToken, resMfaTimePolicy, Errors, BrowserFeatures) {
```
Comment (Ujjwalreddy-okta) [inline test/spec/LoginRouter_spec.js]: Looks like we are not using this in this file.

File: test/helpers/xhr/MFA_REQUIRED_policy_time_based_min.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        "href": "https:\/\/foo.com\/api\/v1\/authn\/cancel",
+        "hints": {
+          "allow": [
+            "POST"
+          ]
+        }
+      }
+    }
+  }
+});
```
Comment (Ujjwalreddy-okta) [inline test/helpers/xhr/MFA_REQUIRED_policy_time_based_min.js]: We can refactor this to include just one mock. Created OKTA-93397 for this.

**Comment by Ujjwalreddy-okta** [comment]
Couple of comments. After that, squash and merge. 🚀

File: src/VerifyDuoController.js (hunk 10 above)
Comment (mauriciocastillosilva-okta) [inline src/VerifyDuoController.js]: That doesn't work, it is undefined. Seems that the only way to get to the appState  is ```this.options.appState```
